subprojects/model-core/src/main/java/org/gradle/internal/reflect/annotations/impl/DefaultTypeAnnotationMetadataStore.java

```diff
@@ -0,0 +1,335 @@
+        Map<Class<? extends Annotation>, ? extends PropertyAnnotationCategory> propertyAnnotationCategories,
+        Collection<Class<?>> ignoredSuperClasses,
+        Collection<Class<?>> ignoreMethodsFromClasses,
+        CrossBuildInMemoryCacheFactory cacheFactory
+    ) {
+        this.recordedTypeAnnotations = ImmutableSet.copyOf(recordedTypeAnnotations);
+        this.propertyAnnotationCategories = ImmutableMap.copyOf(propertyAnnotationCategories);
+        this.cache = cacheFactory.newClassCache();
+        for (Class<?> ignoredSuperClass : ignoredSuperClasses) {
+            cache.put(ignoredSuperClass, EMPTY_TYPE_ANNOTATION_METADATA);
+        }
+        this.ignoredMethods = allMethodsOf(ignoreMethodsFromClasses);
+    }
+
+    private static ImmutableSet<Equivalence.Wrapper<Method>> allMethodsOf(Iterable<Class<?>> classes) {
+        List<Equivalence.Wrapper<Method>> methods = Lists.newArrayList();
+        for (Class<?> clazz : classes) {
+            for (Method method : clazz.getMethods()) {
+                methods.add(SIGNATURE_EQUIVALENCE.wrap(method));
+            }
+        }
+        return ImmutableSet.copyOf(methods);
+    }
+
+    @Override
+    public TypeAnnotationMetadata getTypeAnnotationMetadata(Class<?> type) {
+        return cache.get(type, t -> createTypeAnnotationMetadata(type));
+    }
+
+    private TypeAnnotationMetadata createTypeAnnotationMetadata(Class<?> type) {
+        ImmutableSet.Builder<Annotation> typeAnnotations = ImmutableSet.builder();
+        for (Annotation typeAnnotation : type.getDeclaredAnnotations()) {
+            if (recordedTypeAnnotations.contains(typeAnnotation.annotationType())) {
+                typeAnnotations.add(typeAnnotation);
+            }
+        }
+
+        Map<String, PropertyAnnotationMetadataBuilder> propertyBuilders = extractPropertiesFrom(type);
+        Map<String, PropertyAnnotationMetadata> propertiesMetadata = new HashMap<>();
+        visitSuperTypes(type, superType -> {
+            for (PropertyAnnotationMetadata propertyAnnotationMetadata : superType.getPropertiesAnnotationMetadata()) {
+                String propertyName = propertyAnnotationMetadata.getPropertyName();
+                PropertyAnnotationMetadataBuilder builder = propertyBuilders.get(propertyName);
+                if (builder == null) {
+                    propertiesMetadata.putIfAbsent(propertyName, propertyAnnotationMetadata);
+                } else {
+                    builder.inherit(propertyAnnotationMetadata.getAnnotations());
+                }
+            }
+        });
+        propertyBuilders.forEach((propertyName, builder) -> {
+            propertiesMetadata.put(propertyName, builder.build());
+        });
+        return new DefaultTypeAnnotationMetadata(typeAnnotations.build(), propertiesMetadata.values());
+    }
+
+    private void visitSuperTypes(Class<?> type, Consumer<? super TypeAnnotationMetadata> visitor) {
```

Comment:
I don't like very much that a new way to walk the type hierarchy is introduced here. After `Types.walkTypeHierarchy` and `ClassInspector.visitGraph` I don't think we need an new one. Can't we just use `Types.walkTypeHierarchy` or a variation of it?

Reply:
This doesn't walk the hierarchy, only visits the direct super-types. The advantage here is that instead of always visiting the whole hierarchy anew, we can reuse metadata about super-types. We should check the other usages if they can be converted to this approach.